src/NetAnalyzers/Core/Microsoft.CodeQuality.Analyzers/QualityGuidelines/AvoidMultipleEnumerations/FlowAnalysis/InvocationEntity.cs

```diff
@@ -0,0 +1,40 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the MIT license.  See License.txt in the project root for license information.
+
+using System.Collections.Immutable;
+using Analyzer.Utilities;
+using Microsoft.CodeAnalysis.FlowAnalysis.DataFlow;
+
+namespace Microsoft.CodeQuality.Analyzers.QualityGuidelines.AvoidMultipleEnumerations.FlowAnalysis
+{
+    /// <summary>
+    /// The entity that could be enumerated by for each loop or method.
+    /// </summary>
+    internal class InvocationEntity : CacheBasedEquatable<InvocationEntity>
+    {
+        /// <summary>
+        /// A set of the possible locations of a local or parameter symbol.
+        /// TODO: We should support skip the deferred method call
+        /// e.g. var d = Enumerable.Range(1, 10)
+        ///      d.ToArray();
+        ///      var e = d.Select(i => i)
+        ///      e.ToArray();
+        /// For e, Locations will point to d.Select(i => i), but ideally it should point to the local 'd'
+        /// </summary>
+        private ImmutableHashSet<AbstractLocation> Locations { get; }
```

Comment:
So this is a TODO item,
basically for the example
```
var d = Enumerable.Range(1, 10);
d.ToArray();
 var e = d.Select(i => i)
 e.ToArray();
```
The real enumerated local for `e.ToArray()` is `d`

In my mind the real Location for here is a new union class that could be either 
1. No Creation operation symbol (Parameter symbol)
2. Creation operation

However since this PR is quite big now and I would like to do this in a follow PR.
(If other part this PR looks good then probably we can merge it to a feature branch)

But this